[BugFix] some bug fix for auto increment (#20575)

srlch opened the pull request:
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #20575

## Problem Summary(Required) ：
Problem 1:
Currently, the auto-increment column does not support modifying a column into an AUTO_INCREMENT column. But it is possible that modify an AUTO_INCREMENT column into a normal column. This should be aborted.

Problem 2:
There is no mechanism to handle the overflow of the auto-increment id counter, which will cause be crash.

Problem 3:
The planning for inserting into is not feasible when the auto-increment is not specified and defined before another normal column.

Problem 4:
When we recover a table, the auto-increment id will restart from the beginning, which is a problem.

Solution:
1. Abort the case that we modify an AUTO_INCREMENT column into a normal column.
2. Throw an exception when the id counter is overflowing.
3. Fix the planning problem for the insert statement.
4. remove the auto increment id when truly erasing the table.5. 
5. add some test case for auto increment.
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

imay commented on be/src/storage/storage_engine.cpp:
```diff
@@ -1199,7 +1199,12 @@ Status StorageEngine::get_next_increment_id_interval(int64_t tableid, size_t num
 
         if (!st.ok() || alloc_result.status.status_code != TStatusCode::OK) {
-            return Status::InternalError("auto increment allocate failed");
+            std::string err_msg;
```
need to use std::stringstream to reduce copy

imay commented on be/src/storage/storage_engine.cpp:
```diff
@@ -1199,7 +1199,12 @@ Status StorageEngine::get_next_increment_id_interval(int64_t tableid, size_t num
         if (!st.ok() || alloc_result.status.status_code != TStatusCode::OK) {
-            return Status::InternalError("auto increment allocate failed");
+            std::string err_msg;
+            for (auto msg : alloc_result.status.error_msgs) {
```
```suggestion
            for (auto& msg : alloc_result.status.error_msgs) {
```
use reference to reduce copy

imay commented on fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:
```diff
@@ -380,6 +380,11 @@ private void processModifyColumn(ModifyColumnClause alterClause, OlapTable olapT
 
         Column oriColumn = schemaForFinding.get(modColIndex);
+
+        if (oriColumn.isAutoIncrement()) {
```
should add UT/sql-test to cover this case

imay reviewed (commented):
You need to add UT to cover this fixes.

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:
```diff
@@ -1584,6 +1585,17 @@ public TAllocateAutoIncrementIdResult allocAutoIncrementId(TAllocateAutoIncremen
+            return result;
+        }
+
+        if (nextId == null) {
```
When the nextId == null?

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:
```diff
@@ -308,6 +308,11 @@ protected synchronized void eraseDatabase(long currentTimeMs) {
             RecycleDatabaseInfo dbInfo = entry.getValue();
             Database db = dbInfo.getDb();
             if (canErase(db.getId(), currentTimeMs)) {
+                List<Table> tables = db.getTables();
```
There is no need to delete the table's incrementId for erase db. Because the db's table has been put to idToTableInfo, it will be erased on the eraseTable function.

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:
```diff
@@ -340,6 +345,13 @@ private synchronized void eraseDatabaseWithSameName(String dbName) {
     }
 
     public synchronized void replayEraseDatabase(long dbId) {
+        RecycleDatabaseInfo dbInfo = idToDatabase.get(dbId);
```
same to above.

srlch committed:
[BugFix] some bug fix for auto increment (#20575)

Problem 1:
Currently, auto increment column does not support modify a column into a AUTO_INCREMENT column. But it is possible that modify a AUTO_INCREMENT column into a normal column. This should be abort.

Problem 2:
There is no mechanism to handle the overflow of the auto increment id counter, which will cause be crash.

Problem 3:
The planning for insert into is not feasible when the auto increment is not specified and defined before other normal column.

Problem 4:
When we recover a table, the auto increment id will restart from 0, which is a problem.

Solution:
1. Abort the case that we modify a AUTO_INCREMENT column into a normal column.
2. Throw an exception when the id counter is overflow.
3. Fix the planning problem for insert statement.
4. remove the auto increment id when truly erase the table.
5. add some test case for auto increment column.

Signed-off-by: srlch <linzichao@starrocks.com>

wanpengfei-git commented:
@Mergifyio backport branch-3.0
